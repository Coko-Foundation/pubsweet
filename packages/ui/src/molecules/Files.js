import React from 'react'
import styled from 'styled-components'
import Upload from './Upload'

const Root = styled.div``
const Uploader = styled.div``

const FileList = styled.div`
  display: flex;
  flex-direction: column;
  font-size: 0.9em;
  font-style: italic;
  line-height: 1.5;
`

const AttachButton = styled.button.attrs({
  type: 'button',
})`
  background: transparent;
  border: 1px dashed grey;
  cursor: pointer;
  font-family: inherit;
  font-size: inherit;
  margin-bottom: 2em;
  padding: 10px;
`

/* Not used for now
.button {
  background: transparent;
  border: 1px dashed grey;
  cursor: pointer;
  font-family: inherit;
  font-size: inherit;
  margin-bottom: 2em;
  padding: 10px;
}

.button:hover {
  border-color: var(--color-primary);
  color: var(--color-primary);
}
*/

class Files extends React.Component {
  constructor(props) {
    super(props)

    this.state = {
      values: props.value || [],
      uploads: [],
    }
  }

  handleClick = () => {
    this.fileInput.click()
  }

  handleChange = event => {
    const { uploads } = this.state

    Array.from(event.target.files).forEach(file => {
      uploads.push({
        file,
        request: this.props.uploadFile(file),
      })
    })

    this.setState({ uploads })
  }

  handleUploadedFile = ({ file, url }) => {
    const values = this.state.values.concat({
      name: file.name,
      url,
    })

    const uploads = this.state.uploads.filter(
      item => item.file.name !== file.name,
    )

    this.setState({ values, uploads })

    this.props.onChange(values)
  }

  render() {
    const { name, buttonText, uploadingFile, uploadedFile } = this.props
    const { values, uploads } = this.state

    return (
      <Root>
        <Uploader>
          <AttachButton onClick={() => this.fileInput.click()}>
            {buttonText}
          </AttachButton>

          <input
            multiple
            name={name}
            onChange={this.handleChange}
            ref={input => (this.fileInput = input)}
            style={{ display: 'none' }}
            type="file"
          />
        </Uploader>

        <FileList>
          {uploads &&
            uploads.map(upload => (
              <Upload
                file={upload.file}
                handleUploadedFile={this.handleUploadedFile}
                key={upload.file.name}
                render={uploadingFile}
                request={upload.request}
              />
            ))}

          {values && values.map(uploadedFile)}
        </FileList>
      </Root>
    )
  }
}

export default Files
