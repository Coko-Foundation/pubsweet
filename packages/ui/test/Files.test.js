import React from 'react'
import renderer from 'react-test-renderer'
import 'jest-styled-components'

import Files from '../src/molecules/Files'

const value = [
  {
    name: 'IMG_4346.JPG',
    type: 'JPG',
    size: 4346,
  },
]

describe('Files', () => {
  it('renders correctly', () => {
    const tree = renderer.create(
      <Files
        buttonText="â†‘ Choose a file to upload"
        uploadedFile={value => <div key={value.name}>{value.name}</div>}
        uploadFile={file => new XMLHttpRequest()}
        uploadingFile={({ file, progress, error }) => (
          <div style={{ color: 'gray' }}>{file.name}</div>
        )}
        value={value}
      />,
    )
    expect(tree).toMatchSnapshot()
  })
})
