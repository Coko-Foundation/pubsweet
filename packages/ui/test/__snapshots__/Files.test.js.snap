// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Files renders correctly 1`] = `
Array [
  .c0 {
  background: transparent;
  border: dashed;
  height: calc( * 3);
  cursor: pointer;
  margin-bottom: calc( * 3);
}

<button
    className="c0"
    onClick={[Function]}
    type="button"
  >
    â†‘ Choose a file to upload
  </button>,
  <input
    multiple={true}
    onChange={[Function]}
    style={
      Object {
        "display": "none",
      }
    }
    type="file"
  />,
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: left;
  -webkit-justify-content: left;
  -ms-flex-pack: left;
  justify-content: left;
}

<div
    className="c0"
  />,
]
`;
